File: aznavrail-react/src/__tests__/AzPagesOrdering.test.ts
import { orderBackgrounds, orderOnscreen } from '../AzNavHost';

/** Parity tests for the pages Z-ordering helpers (mirrors the Android AzPagesOrderingTest). */
describe('pages ordering', () => {
  const bg = (id: string, weight: number, page: number) => ({ id, weight, page, content: null });
  const on = (id: string, page: number) => ({ id, page, alignment: 'TopStart' as any, content: null });

  it('orders backgrounds higher-page-furthest-back when pages enabled', () => {
    const items = [bg('a', 0, 0), bg('b', 0, 2), bg('c', 0, 1)];
    expect(orderBackgrounds(items, true).map(i => i.page)).toEqual([2, 1, 0]);
  });

  it('breaks background ties within a page by weight (lower further back)', () => {
    const items = [bg('heavy', 5, 1), bg('light', 0, 1)];
    expect(orderBackgrounds(items, true).map(i => i.weight)).toEqual([0, 5]);
  });

  it('falls back to legacy weight sort when pages disabled', () => {
    const items = [bg('a', 10, 0), bg('b', 1, 5)];
    expect(orderBackgrounds(items, false).map(i => i.weight)).toEqual([1, 10]);
  });

  it('orders onscreen higher-page-furthest-back when pages enabled', () => {
    const items = [on('front', 0), on('back', 3), on('mid', 1)];
    expect(orderOnscreen(items, true).map(i => i.page)).toEqual([3, 1, 0]);
  });

  it('keeps declaration order within a page (stable)', () => {
    const items = [on('first', 1), on('second', 1)];
    expect(orderOnscreen(items, true).map(i => i.id)).toEqual(['first', 'second']);
  });

  it('preserves declaration order when pages disabled', () => {
    const items = [on('a', 9), on('b', 0)];
    expect(orderOnscreen(items, false).map(i => i.id)).toEqual(['a', 'b']);
  });
});

File: aznavrail/src/test/java/com/hereliesaz/aznavrail/AzPagesOrderingTest.kt
package com.hereliesaz.aznavrail

import androidx.compose.ui.Alignment
import org.junit.Assert.assertEquals
import org.junit.Test

/** Unit tests for the pages Z-ordering helpers [azOrderBackgrounds] and [azOrderOnscreen]. */
class AzPagesOrderingTest {

    private fun bg(weight: Int, page: Float) = AzBackgroundItem(weight = weight, page = page) {}
    private fun on(tag: Alignment, page: Float) = AzOnscreenItem(alignment = tag, page = page) {}

    @Test
    fun backgrounds_higherPageDrawnFurtherBack_whenPagesEnabled() {
        val a = bg(weight = 0, page = 0f)
        val b = bg(weight = 0, page = 2f)
        val c = bg(weight = 0, page = 1f)
        // Back-to-front render order: highest page first (backmost), lowest page last (front).
        assertEquals(
            listOf(2f, 1f, 0f),
            azOrderBackgrounds(listOf(a, b, c), pagesEnabled = true).map { it.page },
        )
    }

    @Test
    fun backgrounds_weightBreaksTiesWithinAPage() {
        val heavy = bg(weight = 5, page = 1f)
        val light = bg(weight = 0, page = 1f)
        // Within the same page, lower weight is further back (drawn first).
        assertEquals(
            listOf(0, 5),
            azOrderBackgrounds(listOf(heavy, light), pagesEnabled = true).map { it.weight },
        )
    }

    @Test
    fun backgrounds_fallBackToLegacyWeightSort_whenPagesDisabled() {
        val a = bg(weight = 10, page = 0f)
        val b = bg(weight = 1, page = 5f)
        // Pages off: order by weight ascending, page ignored.
        assertEquals(
            listOf(1, 10),
            azOrderBackgrounds(listOf(a, b), pagesEnabled = false).map { it.weight },
        )
    }

    @Test
    fun onscreen_higherPageDrawnFurtherBack_whenPagesEnabled() {
        val front = on(Alignment.TopStart, page = 0f)
        val back = on(Alignment.TopEnd, page = 3f)
        val mid = on(Alignment.Center, page = 1f)
        assertEquals(
            listOf(3f, 1f, 0f),
            azOrderOnscreen(listOf(front, back, mid), pagesEnabled = true).map { it.page },
        )
    }

    @Test
    fun onscreen_samePageKeepsDeclarationOrder() {
        val first = on(Alignment.TopStart, page = 1f)
        val second = on(Alignment.TopEnd, page = 1f)
        // Stable sort: same page keeps declaration order.
        assertEquals(
            listOf(Alignment.TopStart, Alignment.TopEnd),
            azOrderOnscreen(listOf(first, second), pagesEnabled = true).map { it.alignment },
        )
    }

    @Test
    fun onscreen_preservesDeclarationOrder_whenPagesDisabled() {
        val a = on(Alignment.TopStart, page = 9f)
        val b = on(Alignment.BottomEnd, page = 0f)
        // Pages off: declaration order unchanged, page ignored.
        assertEquals(
            listOf(Alignment.TopStart, Alignment.BottomEnd),
            azOrderOnscreen(listOf(a, b), pagesEnabled = false).map { it.alignment },
        )
    }
}
